controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java

```diff
@@ -0,0 +1,104 @@
+ * interceptor mechanism.
+ * In case of REST calls, this class routes the call to specific PravegaAuthHandler.
+ */
+@Slf4j
+public class PravegaAuthManager {
+    private final GRPCServerConfig serverConfig;
+    @GuardedBy("this")
+    private final Map<String, PravegaAuthHandler> handlerMap;
+
+    public PravegaAuthManager(GRPCServerConfig serverConfig) {
+        this.serverConfig = serverConfig;
+        this.handlerMap = new HashMap<>();
+    }
+
+    private PravegaAuthHandler getHandler(String handlerName) throws PravegaAuthenticationException {
+        PravegaAuthHandler retVal;
+        synchronized (this) {
+                retVal = handlerMap.get(handlerName);
+            }
+            if (retVal == null) {
+            throw new PravegaAuthenticationException("Handler does not exist for method " + handlerName);
+        }
+        return retVal;
+    }
+
+    /**
+     * API to authenticate and authroize access to a given resource.
+     * @param resource The resource identifier for which the access needs to be controlled.
+     * @param headers  Custom headers used for authentication.
+     * @param level    Expected level of access.
+     * @return         Returns true if the entity represented by the custom auth headers had given level of access to the resource.
+     * @throws PravegaAuthenticationException Exception faced during authentication/authorization.
+     */
+    public boolean authenticate(String resource, MultivaluedMap<String, String> headers, PravegaAuthHandler.PravegaAccessControlEnum level) throws PravegaAuthenticationException {
+        boolean retVal = false;
+        try {
+            Map<String, String> paramMap = headers.entrySet().stream().collect(Collectors.toMap(k -> k.getKey(), k -> k.getValue().get(0)));
+            String method = paramMap.get("method");
+
+            PravegaAuthHandler handler = getHandler(method);
+            retVal = handler.authenticate(paramMap) &&
+                    handler.authorize(resource, paramMap).ordinal() >= level.ordinal();
+        } catch (RuntimeException e) {
+            throw new PravegaAuthenticationException(e);
+        }
+        return retVal;
+    }
+
+    /**
+     * Loads the custom implementations of the PravegaAuthHandler interface dynamically. Registers the interceptors with grpc.
+     * Stores the implementation in a local map for routing the REST auth request.
+     * @param builder The grpc service builder to register the interceptors.
+     */
+    public void registerInterceptors(ServerBuilder<?> builder) {
+        try {
+            if (serverConfig.isAuthorizationEnabled()) {
+                ServiceLoader<PravegaAuthHandler> loader = ServiceLoader.load(PravegaAuthHandler.class);
+                for (PravegaAuthHandler handler : loader) {
+                    try {
+                        handler.initialize(serverConfig);
```

Comment:
Here, we pass the serverConfig to all the handlers.   It is the serverConfig that contains stock Pravega settings like whether or not auth is enabled, tls, path to the passwordfile ...

For 3rd party implementors of the handler interface, they may need their own version of the serverConfig object.   For example, some OpenID connect implementation may need the IP of the identity provider etc.

What mechanism can 3rd party implementors use to provide custom settings to their plugin?  Today, it seems that could not use the initialize() call for this.  But may need to use system properties or env variables to pass those settings?

Reply:
The way configuration is implemented and passed on to the plugin is left to the plugin implementations. This is just in case if the implementation needs to know more about the controller.